rfc/rfc-89/rfc-89.md

```diff
@@ -0,0 +1,423 @@
+
+*Note:* Before updating the hashing_config, users must manually stop all ingestion jobs.
+
+```text
+Firstly DDL phase creates initial version 
+ls .hoodie/.hashing_meta/
+.hoodie/.hashing_meta/00000000000000000.hashing_config
+{
+   "rule": "rule-engine",
+   "expressions": "expression1,bucket-number1;expression2,bucket-number2",
+   "default_bucket_number": "default-bucket-number"
+}
+
+Secondly call PartitionBucketIndexManager(overwrite => 'expression3,bucket-number3;expression1,bucket-number1;expression2,bucket-number2', 'dry-run' = 'false')
+ls .hoodie/.hashing_meta/
+.hoodie/.hashing_meta/00000000000000000.hashing_config (deprecated)
+.hoodie/.hashing_meta/20250303095546020.hashing_config (used)
+{
+   "rule": "rule-engine",
+   "expressions": "expression3,bucket-number3;expression1,bucket-number1;expression2,bucket-number2",
+   "default_bucket_number": "default-bucket-number"
+}
+```
+
+The expression written at the front has a higher priority, and the expression has a higher priority than the default_bucket_number.
+
+About Hashing_Config Rollback And Cleanup
+
+The Bucket Rescale operation is essentially an insert overwrite action and can seamlessly integrate with Hudi's rollback mechanism. 
+
+This requires integration with Hudi's existing rollback mechanism, such that when rolling back an INSERT OVERWRITE operation, 
+the corresponding hashing_config file should be deleted if existed.
+
+
+About Hashing_Config Archive
+
+As the number of times users modify expressions increases, the corresponding number of hashing_config files also grows. 
+To address this, we need a mechanism to archive active hashing_config files while 
+preserving all historical hashing_config information in an archive file `archived.hashing_config` under `.hoodie/.hashing_meta/archive/` 
+to support scenarios like time travel.
+
+A new archived hashing config file `.hoodie/.hashing_meta/archive/archived.hashing_config`, will be created. Its schema is as follows:
+```json
+{
+  "<instant1>" : {
+     "rule": "rule-engine",
+     "expressions": "expression",
+     "default_bucket_number": "number"
+  },
+   "<instant2>" : {
+      "rule": "rule-engine",
+      "expressions": "expression",
+      "default_bucket_number": "number"
+   }
+}
+```
+
+The archive logic is triggered when invoking `call partitionBucketIndexManager(overwrite => 'new-expressions', dry-run => 'false')` to modify expressions. 
+
+Once the number of active hashing configs exceeds 5, the oldest expression’s metadata is automatically merged into `archived.hashing_config`. 
+The workflow is as follows:
+
+![archive.jpg](archive.jpg)
+
+**we always create a new temp file and rename it to archived.hashing_config**
+
+About Loading Hashing_Config
+
+First, attempt to retrieve the latest `<replace-commit>.hashing_config` from the committed replace active timeline.
+If it does not exist, fetch the latest hashing_config from `instants before active timeline start`.
```

Comment:
Should we load from the archived.hashing_config?

Reply:
We always keep at least one valid active hashing_config after a hashing_config archived. This means the hashing_config time order is:
`archived hashing_config < valid hashing_config < active timeline start < active hashing_config`
Therefore, when retrieving the latest hashing_config, we don't need to load the archived.hashing_config.